models/eprop_iaf.cpp

```diff
@@ -0,0 +1,442 @@
+  , gamma_( 0.3 )
+  , I_e_( 0.0 )
+  , surrogate_gradient_function_( "piecewise_linear" )
+  , t_ref_( 2.0 )
+  , tau_m_( 10.0 )
+  , V_min_( -std::numeric_limits< double >::max() )
+  , V_th_( -55.0 - E_L_ )
+  , kappa_( 0.97 )
+  , kappa_reg_( 0.97 )
+  , eprop_isi_trace_cutoff_( 1000.0 )
+{
+}
+
+eprop_iaf::State_::State_()
+  : learning_signal_( 0.0 )
+  , r_( 0 )
+  , surrogate_gradient_( 0.0 )
+  , i_in_( 0.0 )
+  , v_m_( 0.0 )
+  , z_( 0.0 )
+  , z_in_( 0.0 )
+{
+}
+
+eprop_iaf::Buffers_::Buffers_( eprop_iaf& n )
+  : logger_( n )
+{
+}
+
+eprop_iaf::Buffers_::Buffers_( const Buffers_&, eprop_iaf& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Getter and setter functions for parameters and state
+ * ---------------------------------------------------------------- */
+
+void
+eprop_iaf::Parameters_::get( DictionaryDatum& d ) const
+{
+  def< double >( d, names::C_m, C_m_ );
+  def< double >( d, names::c_reg, c_reg_ );
+  def< double >( d, names::E_L, E_L_ );
+  def< double >( d, names::f_target, f_target_ );
+  def< double >( d, names::beta, beta_ );
+  def< double >( d, names::gamma, gamma_ );
+  def< double >( d, names::I_e, I_e_ );
+  def< std::string >( d, names::surrogate_gradient_function, surrogate_gradient_function_ );
+  def< double >( d, names::t_ref, t_ref_ );
+  def< double >( d, names::tau_m, tau_m_ );
+  def< double >( d, names::V_min, V_min_ + E_L_ );
+  def< double >( d, names::V_th, V_th_ + E_L_ );
+  def< double >( d, names::kappa, kappa_ );
+  def< double >( d, names::kappa_reg, kappa_reg_ );
+  def< double >( d, names::eprop_isi_trace_cutoff, eprop_isi_trace_cutoff_ );
+}
+
+double
+eprop_iaf::Parameters_::set( const DictionaryDatum& d, Node* node )
+{
+  // if leak potential is changed, adjust all variables defined relative to it
+  const double ELold = E_L_;
+  updateValueParam< double >( d, names::E_L, E_L_, node );
+  const double delta_EL = E_L_ - ELold;
+
+  V_th_ -= updateValueParam< double >( d, names::V_th, V_th_, node ) ? E_L_ : delta_EL;
+  V_min_ -= updateValueParam< double >( d, names::V_min, V_min_, node ) ? E_L_ : delta_EL;
+
+  updateValueParam< double >( d, names::C_m, C_m_, node );
+  updateValueParam< double >( d, names::c_reg, c_reg_, node );
+
+  if ( updateValueParam< double >( d, names::f_target, f_target_, node ) )
+  {
+    f_target_ /= 1000.0; // convert from spikes/s to spikes/ms
+  }
+
+  updateValueParam< double >( d, names::beta, beta_, node );
+  updateValueParam< double >( d, names::gamma, gamma_, node );
+  updateValueParam< double >( d, names::I_e, I_e_, node );
+
+  if ( updateValueParam< std::string >( d, names::surrogate_gradient_function, surrogate_gradient_function_, node ) )
+  {
+    eprop_iaf* nrn = dynamic_cast< eprop_iaf* >( node );
+    assert( nrn );
+    auto compute_surrogate_gradient = nrn->find_surrogate_gradient( surrogate_gradient_function_ );
+    nrn->compute_surrogate_gradient_ = compute_surrogate_gradient;
```

Comment:
@JesusEV This looks good, but I don't see the point of assigning to a local variable first, so I suggest
```suggestion
    nrn->compute_surrogate_gradient_ = nrn->find_surrogate_gradient( surrogate_gradient_function_ );
```

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/3207/commits/5b8f25c2565f90516b8fe5ac592ca48bb28445fc